File: configuration-manager-go-client/rules_service.go
package configmanager

import (
	"context"
	"fmt"
	"net/url"
	"strconv"
)

// GetAll retrieves rules for a configuration
func (r *RulesService) GetAll(ctx context.Context, options *GetRulesOptions) (*RulesListResponse, error) {
	if options == nil || options.ConfigurationID == "" {
		return nil, NewValidationError("configuration ID is required", 0)
	}

	params := url.Values{}
	params.Set("configurationId", options.ConfigurationID)

	resp, err := r.client.GetWithQuery(ctx, "/rules", params)
	if err != nil {
		return nil, fmt.Errorf("get rules request failed: %w", err)
	}
	defer r.client.CloseResponse(resp)

	var response RulesListResponse
	if err := r.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode rules response: %w", err)
	}

	return &response, nil
}

// Create creates a new rule
func (r *RulesService) Create(ctx context.Context, request *CreateRuleRequest) (*RuleResponse, error) {
	if request == nil {
		return nil, NewValidationError("create rule request cannot be nil", 0)
	}

	if request.ConfigurationID == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	if request.PropertyPath == "" {
		return nil, NewValidationError("property path cannot be empty", 0)
	}

	resp, err := r.client.Post(ctx, "/rules", request)
	if err != nil {
		return nil, fmt.Errorf("create rule request failed: %w", err)
	}
	defer r.client.CloseResponse(resp)

	var response RuleResponse
	if err := r.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode create rule response: %w", err)
	}

	return &response, nil
}

// Get retrieves a specific rule by ID
func (r *RulesService) Get(ctx context.Context, id string) (*RuleResponse, error) {
	if id == "" {
		return nil, NewValidationError("rule ID cannot be empty", 0)
	}

	path := fmt.Sprintf("/rules/%s", id)
	resp, err := r.client.Get(ctx, path)
	if err != nil {
		return nil, fmt.Errorf("get rule request failed: %w", err)
	}
	defer r.client.CloseResponse(resp)

	var response RuleResponse
	if err := r.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode rule response: %w", err)
	}

	return &response, nil
}

// Update updates an existing rule
func (r *RulesService) Update(ctx context.Context, id string, request *UpdateRuleRequest) (*RuleResponse, error) {
	if id == "" {
		return nil, NewValidationError("rule ID cannot be empty", 0)
	}

	if request == nil {
		return nil, NewValidationError("update rule request cannot be nil", 0)
	}

	path := fmt.Sprintf("/rules/%s", id)
	resp, err := r.client.Put(ctx, path, request)
	if err != nil {
		return nil, fmt.Errorf("update rule request failed: %w", err)
	}
	defer r.client.CloseResponse(resp)

	var response RuleResponse
	if err := r.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode update rule response: %w", err)
	}

	return &response, nil
}

// Delete deletes a rule
func (r *RulesService) Delete(ctx context.Context, id string) error {
	if id == "" {
		return NewValidationError("rule ID cannot be empty", 0)
	}

	path := fmt.Sprintf("/rules/%s", id)
	resp, err := r.client.Delete(ctx, path)
	if err != nil {
		return fmt.Errorf("delete rule request failed: %w", err)
	}
	defer r.client.CloseResponse(resp)

	return nil
}

// Validate validates a value against rules
func (r *RulesService) Validate(ctx context.Context, request *ValidateRuleRequest) (*RuleValidationResponse, error) {
	if request == nil {
		return nil, NewValidationError("validate rule request cannot be nil", 0)
	}

	if request.ConfigurationID == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	if request.PropertyPath == "" {
		return nil, NewValidationError("property path cannot be empty", 0)
	}

	resp, err := r.client.Post(ctx, "/rules/validate", request)
	if err != nil {
		return nil, fmt.Errorf("validate rule request failed: %w", err)
	}
	defer r.client.CloseResponse(resp)

	var response RuleValidationResponse
	if err := r.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode validate rule response: %w", err)
	}

	return &response, nil
}

// GetForPath gets rules for a specific configuration and path
func (r *RulesService) GetForPath(ctx context.Context, configID, path string, options *GetRulesForPathOptions) (*RulesListResponse, error) {
	if configID == "" {
		return nil, NewValidationError("configuration ID cannot be empty", 0)
	}

	if path == "" {
		return nil, NewValidationError("path cannot be empty", 0)
	}

	// URL encode the path
	encodedPath := url.PathEscape(path)
	endpoint := fmt.Sprintf("/rules/configuration/%s/path/%s", configID, encodedPath)

	params := url.Values{}
	if options != nil && options.IncludeInherited != nil && !*options.IncludeInherited {
		params.Set("includeInherited", "false")
	}

	resp, err := r.client.GetWithQuery(ctx, endpoint, params)
	if err != nil {
		return nil, fmt.Errorf("get rules for path request failed: %w", err)
	}
	defer r.client.CloseResponse(resp)

	var response RulesListResponse
	if err := r.client.DecodeJSON(resp, &response); err != nil {
		return nil, fmt.Errorf("failed to decode rules for path response: %w", err)
	}

	return &response, nil
}

File: configuration-manager-python-client/configuration_manager_client/services/rules_service.py
"""Rules management service for Configuration Manager API."""

from typing import Optional
from urllib.parse import quote

from ..http_client import HTTPClient
from ..models.rules import (
    GetRulesOptions,
    RulesListResponse,
    CreateRuleRequest,
    UpdateRuleRequest,
    RuleResponse,
    ValidateRuleRequest,
    RuleValidationResponse,
    GetRulesForPathOptions,
)
from ..exceptions import ValidationError
from ..utils import from_dict


class RulesService:
    """Rules management service client."""

    def __init__(self, http_client: HTTPClient):
        """Initialize rules service.
        
        Args:
            http_client: HTTP client instance
        """
        self.client = http_client

    def get_all(self, options: GetRulesOptions) -> RulesListResponse:
        """Get rules for a configuration.
        
        Args:
            options: Options containing configuration ID
            
        Returns:
            RulesListResponse containing list of rules
            
        Raises:
            ValidationError: If configuration_id is empty
        """
        if not options.configuration_id:
            raise ValidationError("Configuration ID is required")

        params = {"configurationId": options.configuration_id}
        response = self.client.get("/rules", params=params)
        return from_dict(RulesListResponse, response.json())

    def create(self, request: CreateRuleRequest) -> RuleResponse:
        """Create a new rule.
        
        Args:
            request: Create rule request
            
        Returns:
            RuleResponse containing created rule
            
        Raises:
            ValidationError: If request is invalid
        """
        if not request.configuration_id:
            raise ValidationError("Configuration ID cannot be empty")

        if not request.property_path:
            raise ValidationError("Property path cannot be empty")

        data = {
            "configurationId": request.configuration_id,
            "propertyPath": request.property_path,
            "ruleType": request.rule_type.value,
            "ruleConfig": request.rule_config,
        }

        if request.error_message:
            data["errorMessage"] = request.error_message

        if request.enabled is not None:
            data["enabled"] = request.enabled

        response = self.client.post("/rules", json_data=data)
        return from_dict(RuleResponse, response.json())

    def get(self, rule_id: str) -> RuleResponse:
        """Get a specific rule by ID.
        
        Args:
            rule_id: Rule ID
            
        Returns:
            RuleResponse containing rule information
            
        Raises:
            ValidationError: If rule_id is empty
        """
        if not rule_id:
            raise ValidationError("Rule ID cannot be empty")

        response = self.client.get(f"/rules/{rule_id}")
        return from_dict(RuleResponse, response.json())

    def update(self, rule_id: str, request: UpdateRuleRequest) -> RuleResponse:
        """Update an existing rule.
        
        Args:
            rule_id: Rule ID
            request: Update rule request
            
        Returns:
            RuleResponse containing updated rule
            
        Raises:
            ValidationError: If rule_id is empty
        """
        if not rule_id:
            raise ValidationError("Rule ID cannot be empty")

        data = {}

        if request.property_path:
            data["propertyPath"] = request.property_path

        if request.rule_type:
            data["ruleType"] = request.rule_type.value

        if request.rule_config is not None:
            data["ruleConfig"] = request.rule_config

        if request.error_message is not None:
            data["errorMessage"] = request.error_message

        if request.enabled is not None:
            data["enabled"] = request.enabled

        response = self.client.put(f"/rules/{rule_id}", json_data=data)
        return from_dict(RuleResponse, response.json())

    def delete(self, rule_id: str) -> None:
        """Delete a rule.
        
        Args:
            rule_id: Rule ID
            
        Raises:
            ValidationError: If rule_id is empty
        """
        if not rule_id:
            raise ValidationError("Rule ID cannot be empty")

        self.client.delete(f"/rules/{rule_id}")

    def validate(self, request: ValidateRuleRequest) -> RuleValidationResponse:
        """Validate a value against rules.
        
        Args:
            request: Validate rule request
            
        Returns:
            RuleValidationResponse containing validation result
            
        Raises:
            ValidationError: If request is invalid
        """
        if not request.configuration_id:
            raise ValidationError("Configuration ID cannot be empty")

        if not request.property_path:
            raise ValidationError("Property path cannot be empty")

        data = {
            "configurationId": request.configuration_id,
            "propertyPath": request.property_path,
            "value": request.value,
        }

        response = self.client.post("/rules/validate", json_data=data)
        return from_dict(RuleValidationResponse, response.json())

    def get_for_path(
        self,
        config_id: str,
        path: str,
        options: Optional[GetRulesForPathOptions] = None,
    ) -> RulesListResponse:
        """Get rules for a specific configuration and path.
        
        Args:
            config_id: Configuration ID
            path: Property path
            options: Options for getting rules
            
        Returns:
            RulesListResponse containing rules for the path
            
        Raises:
            ValidationError: If parameters are invalid
        """
        if not config_id:
            raise ValidationError("Configuration ID cannot be empty")

        if not path:
            raise ValidationError("Path cannot be empty")

        # URL encode the path
        encoded_path = quote(path, safe='')
        endpoint = f"/rules/configuration/{config_id}/path/{encoded_path}"

        params = {}
        if options and options.include_inherited is not None:
            params["includeInherited"] = "true" if options.include_inherited else "false"

        response = self.client.get(endpoint, params=params)
        return from_dict(RulesListResponse, response.json())
